sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotShiftByZeroOrIntSize.cs

```diff
@@ -1,151 +1,238 @@
+            }
+
+            if (!TryGetConstantValue(tuple.Item2, out var shiftByCount))
+            {
+                return new ShiftInstance(node);
+            }
+
+            var typeSymbol = semanticModel.GetTypeInfo(node).ConvertedType;
+            if (typeSymbol == null)
+            {
+                return new ShiftInstance(node);
+            }
+
+            var variableBitLength = FindTypeSizeOrDefault(typeSymbol);
+            if (variableBitLength == 0)
+            {
+                return new ShiftInstance(node);
+            }
+
+            var issueDescription = FindProblemDescription(variableBitLength, shiftByCount, tuple.Item1, out bool isLiteralZero);
+            return issueDescription == null ? new ShiftInstance(node) : new ShiftInstance(issueDescription, isLiteralZero, node);
+        }
+
+        private static int FindTypeSizeOrDefault(ITypeSymbol typeSymbol)
+        {
+            return mapKnownTypesToIntegerBitSize
+                .Where(kv => typeSymbol.Is(kv.Key))
+                .Select(kv => kv.Value)
+                .FirstOrDefault();
+        }
+
+        private static string FindProblemDescription(int typeSizeInBits, int shiftBy, bool isLeftShift, out bool isLiteralZero)
+        {
+            isLiteralZero = false;
+            if (shiftBy == 0)
+            {
+                isLiteralZero = true;
+                return string.Format(MessageFormat_UselessShift, 0);
+            }
+
+            if (shiftBy < typeSizeInBits)
+            {
+                return null;
+            }
+
+            if (!isLeftShift)
+            {
+                return string.Format(MessageFormat_RightShiftTooLarge, shiftBy);
+            }
+
+            var shiftSuggestion = shiftBy % typeSizeInBits;
+
+            if (typeSizeInBits == 64)
+            {
+                return shiftSuggestion == 0
+                    ? string.Format(MessageFormat_UselessShift, shiftBy)
+                    : string.Format(MessageFormat_ShiftTooLarge, shiftSuggestion);
+            }
+
+            if (shiftSuggestion == 0)
+            {
+                return string.Format(MessageFormat_UseLargerTypeOrPromote,
+                    "less than " + typeSizeInBits);
+            }
+
+            return string.Format(MessageFormat_UseLargerTypeOrPromote, shiftSuggestion);
+        }
+
+        private class ShiftInstance
+        {
+            public Diagnostic Diagnostic { get; }
+            public bool IsLiteralZero { get; }
+            public int Line { get; }
+
+            public ShiftInstance(SyntaxNode node)
+            {
+                Line = node.GetLineNumberToReport();
+            }
+
+            public ShiftInstance(string description, bool isLieralZero, SyntaxNode node)
+                : this(node)
+            {
+                Diagnostic = Diagnostic.Create(rule, node.GetLocation(), description);
+                IsLiteralZero = isLieralZero;
+            }
+        }
+    }
+}

```

Comment:
That's weird that you don't have the extra empty line at the end of the class.

Reply:
Weird indeed. I can see it in both files in VS...